Subject: Remove the VOID_PTR facilitator macro

I've been wondering about the intended use of this facilitator macro
for some time and concluded that it is plainly wrong. It should
take a pointer value from a buffer, but what is the use of a pointer
that pointed to something on the _target_ machine? It cannot be
meaningfully dereferenced on the host.

Okay, pointer arithmetics is done in one place, but IMO it is better
to explicitly state the object size there, because that's how things
are done elsewhere in the code, so it is less surprising for anybody
reading the code.

The irony is underlined by the fact that all three uses of the macro
need a subsequent typecast to ulong, which proves the whole concept
flawed.

Signed-off-by: Petr Tesarik <ptesarik@suse.cz>

---
 defs.h   |    1 -
 memory.c |   27 +++++++++++++++------------
 2 files changed, 15 insertions(+), 13 deletions(-)

--- a/defs.h
+++ b/defs.h
@@ -1787,7 +1787,6 @@ struct builtin_debug_table {
 #define ULONG_PTR(ADDR) *((ulong **)((char *)(ADDR)))
 #define USHORT(ADDR)    *((ushort *)((char *)(ADDR)))
 #define SHORT(ADDR)     *((short *)((char *)(ADDR)))
-#define VOID_PTR(ADDR)  *((void **)((char *)(ADDR)))
 
 struct node_table {
 	int node_id;
--- a/memory.c
+++ b/memory.c
@@ -27,8 +27,8 @@ struct meminfo {           /* general pu
         ulong c_offset;
 	ulong c_num;
 	ulong s_mem; 
-	void *s_freep; 
-	ulong *s_index; 
+	ulong s_freep;
+	ulong s_index;
 	ulong s_inuse;
 	ulong cpucached_cache;
 	ulong cpucached_slab;
@@ -9769,9 +9769,9 @@ dump_slab(struct meminfo *si)
 			return;
         }
 
-	si->s_freep = VOID_PTR(si->slab_buf + OFFSET(kmem_slab_s_s_freep));
+	si->s_freep = ULONG(si->slab_buf + OFFSET(kmem_slab_s_s_freep));
 	si->s_inuse = ULONG(si->slab_buf + OFFSET(kmem_slab_s_s_inuse));
-	si->s_index = ULONG_PTR(si->slab_buf + OFFSET(kmem_slab_s_s_index));
+	si->s_index = ULONG(si->slab_buf + OFFSET(kmem_slab_s_s_index));
 
 	if (!(si->flags & ADDRESS_SPECIFIED)) {
 		fprintf(fp, slab_hdr);
@@ -9898,7 +9898,7 @@ dump_slab_percpu_v2(struct meminfo *si)
 static void
 gather_slab_free_list(struct meminfo *si)
 {
-	ulong *next, obj;
+	ulong next, obj;
 	ulong expected, cnt;
 
 	BNEG(si->addrlist, sizeof(ulong) * (si->c_num+1));
@@ -9933,16 +9933,18 @@ gather_slab_free_list(struct meminfo *si
 		 */
 
 	        if (si->c_flags & SLAB_CFLGS_BUFCTL) 
-                	obj = si->s_mem + ((next - si->s_index) * si->c_offset);
+                	obj = si->s_mem +
+				((next - si->s_index) / sizeof(ulong)
+				 * si->c_offset);
 		else 
-			obj = (ulong)next - si->c_offset;
+			obj = next - si->c_offset;
 
 		si->addrlist[cnt] = obj; 
 
 		if (si->flags & ADDRESS_SPECIFIED) {
 			if (INSLAB(next, si) && 
-		            (si->spec_addr >= (ulong)next) &&
-			    (si->spec_addr < (ulong)(next + 1))) {
+		            (si->spec_addr >= next) &&
+			    (si->spec_addr < next + sizeof(ulong))) {
 				si->found = KMEM_BUFCTL_ADDR;
 				return;
 			}
@@ -9957,7 +9959,7 @@ gather_slab_free_list(struct meminfo *si
 			si->errors++;
 		}
 
-        	readmem((ulong)next, KVADDR, &next, sizeof(void *),
+        	readmem(next, KVADDR, &next, sizeof(ulong),
                 	"s_freep chain entry", FAULT_ON_ERROR);
 	} while (next); 
 
@@ -10105,7 +10107,7 @@ static void
 dump_slab_objects(struct meminfo *si)
 {
 	int i, j;
-	ulong *next;
+	ulong next;
 	int on_free_list; 
 	ulong cnt, expected;
 	ulong bufctl, obj;
@@ -10136,7 +10138,8 @@ dump_slab_objects(struct meminfo *si)
         if (si->c_flags & SLAB_CFLGS_BUFCTL) {
 		for (i = 0, next = si->s_index; i < si->c_num; i++, next++) {
                 	obj = si->s_mem + 
-				((next - si->s_index) * si->c_offset);
+				((next - si->s_index) / sizeof(ulong)
+				 * si->c_offset);
 			DUMP_SLAB_OBJECT();
 		}
 	} else {
